### DISCUSSION TITLE
How to create an InstanceKeeper

### DISCUSSION THREAD

**Discussion by manosbatsis** [arkivanov/Essenty#58] [Q&A] [answered]
In our multiplatform's **Android** module we create an InstanceKeeper singleton for DI (kodein) as follows:

```kotlin
import androidx.lifecycle.ViewModelStore
import com.arkivanov.essenty.instancekeeper.InstanceKeeper
//...

bind<InstanceKeeper>() with singleton { InstanceKeeper(ViewModelStore()) }
```

My questions are:

1) Is it actually necessary to explicitly create and reuse such an instance?
2) If true, what is the equivalent line for my desktop (jvm) module?

**Comment by arkivanov** [comment, marked as answer]
It depends on how you use the InstanceKeeper. Typically it is used to survive configuration changes on Android, and yet use it in common code. So the answers are:

1. On Android, it's better to create `InstanceKeeper` using a `ViewModelStore` from an Activity or a Fragment. So `InstanceKeeper` shouldn't be a singleton, it's supposed to be destroyed when the activity/fragment is finished.
2. On other platforms, the `InstanceKeeper` should be usually scoped to a screen. You can create and pass `InstanceKeeperDispatcher` there.

**Reply by arkivanov** [reply]
Also, navigation libraries can manage `InstanceKeeper` for you. E.g. [Decompose](https://github.com/arkivanov/Decompose) automatically provides an instance of `InstanceKeeper` for every screen.

**Reply by manosbatsis** [reply]
Thanks, we also use decompose, will do some refactoring to pickup  `InstanceKeeper` from there

**Reply by arkivanov** [reply]
In this case please check out the doc - https://arkivanov.github.io/Decompose/component/instance-retaining/
